File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        def dfs(node):
            if not node:
                nodesPreorder.append("null")
                return

            nodesPreorder.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        
        nodesPreorder = []
        dfs(root)
        string = ','.join(nodesPreorder)
        return string
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        def dfs():
            val = next(vals)
            if val == 'null':
                return None
            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node

        vals = iter(data.split(','))
        return dfs()

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    long long prev = LLONG_MIN;

public:
    bool isValidBST(TreeNode* root) {

        return inOrder(root);

        
    }

private:
    bool inOrder(TreeNode* node) {
        if (!node) return true;

        if (!inOrder(node->left)) return false;

        if (node->val <= prev) return false;
        prev = node->val;

        return inOrder(node->right);

    }
};
